### PR TITLE
Re-apply "Make all CF types Equatable and Hashable."

### PR DESCRIPTION
**Pull request by jrose-apple**
This re-applies #4394, reverted in #4435, now that John has fixed CF casting logic in the optimizer. The round-trip casting test is also passing now.

https://bugs.swift.org/browse/SR-2388

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@DougGregor, @parkera, this is technically a source-breaking change since someone may have manually added Hashable or Equatable to a CF type. Then again, that's both a type and protocol they don't own, which isn't really safe because *anyone* could do it ("what if two people did this?").  Do you think we can commit this or do we have to be more circumspect?

**Comment by lattner** [comment]
This seems like acceptable risk to me.

File: test/Interpreter/SDK/cf_without_foundation.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
+// RUN: %target-run-simple-swift | %FileCheck %s
+
+// REQUIRES: executable_test
+// REQUIRES: objc_interop
```
Comment (karwa) [inline test/Interpreter/SDK/cf_without_foundation.swift]: Technically only requires a platform where CF is part of the SDK, doesn't it?

What do you think about moving/copying this test in to Foundation and running it on all platforms?

**Comment by jrose-apple** [comment]
I've found at least one case on GitHub where someone is doing exactly this on CFString, so it's definitely a source-breaking change. May have to wait until Swift 4.

**Review by thomasbarwanitz — COMMENTED** [review]
Thank you

**Review by thomasbarwanitz — COMMENTED** [review]
Thhhhxzzzzzzzzzz

**Comment by jrose-apple** [comment]
Resurrecting this. It's almost certainly breaking, and we don't have a way to conditionalize it on Swift 3/4, but it is an improvement. What do people think?

**Comment by dabrahams** [comment]
Even if this breaks somebody, it's really simple to fix.  +1

**Comment by DougGregor** [comment]
I'm definitely +1 on doing this

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
I still want to do a full test on Apple platforms but there's an error in one test right now.

@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
@swift-ci Please test

**Comment by jrose-apple** [comment]
@akyrtzi, we're now running into the old issue where `-include` doesn't actually count as having included something (rdar://problem/22083824). Are you okay with the change to test/IDE/print_clang_header.swift to kick that can down the road a little further?

**Comment by jrose-apple** [comment]
@swift-ci Please test macOS

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test Linux

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please test source compatibility

**Comment by jrose-apple** [comment]
@swift-ci Please smoke test

**Pull request merged**